Updated guide to the new format (#2)
hazelcast-guides/caching-micronaut-microservices-on-kubernetes: README.adoc

:github-address: https://github.com/hazelcast-guides/caching-micronaut-microservices-on-kubernetes
:templates-url: https://raw.githubusercontent.com/hazelcast-guides/adoc-templates/master
:hazelcast: Hazelcast IMDG
:framework: Micronaut

= Getting Started with Hazelcast using Micronaut in Kubernetes

This guide helps you start with Hazelcast in Micronaut based Microservice and deploy into Kubernetes

https://raw.githubusercontent.com/hazelcast-guides/adoc-templates/master/link-to-repo.adoc

== What You’ll Learn

In this guide, you will deploy Micronaut application into a Kubernetes cluster. Hazelcast cluster members will automatically discover each other using https://github.com/hazelcast/hazelcast-kubernetes[Hazelcast Kubernetes discovery plugin].

== Prerequisites

-   ~15 minutes

-   Docker (Docker for Desktop is good enough)

-   Kubernetes cluster (Docker for Desktop or Minikube is good enough)

-  JDK 1.8+

-  Apache Maven 3.2+

== About Deploying the Application to Kubernetes

> We will take the application that we built on https://github.com/hazelcast-guides/hazelcast-embedded-micronaut[Hazelcast Embedded Micronaut] and deploy it on Kubernetes. https://github.com/hazelcast-guides/caching-micronaut-microservices-on-kubernetes[Caching Micronaut Microservices on kubernetes] has the updated version of the application.

Author: That guide doesn't have the kubernetes dependencies applied. I may have used incorrect wording. But I think we should have a link to that guide at the start of this one.

> We will take the application that we built on https://github.com/hazelcast-guides/hazelcast-embedded-micronaut[Hazelcast Embedded Micronaut] and deploy it on Kubernetes. https://github.com/hazelcast-guides/caching-micronaut-microservices-on-kubernetes[Caching Micronaut Microservices on kubernetes] has the updated version of the application.

Reviewer: Then we probably should say "In this guide, we deploy the app we built on < hazelcast-embedded-micronaut > on k8s env." or something like that. Otherwise it may be misleading such that there are 2 other apps we built other than this.

In `pom.xml` we add the following dependecy for cluster members to find each other in the cluster.
[source]
----
<dependency>
    <groupId>com.hazelcast</groupId>
    <artifactId>hazelcast-kubernetes</artifactId>
    <version>${hazelcast-kubernetes.version}</version>
</dependency>
----
And in the code we change the `HazelcastAdditionalSettings` class as follows:

[source,java]
----
@Singleton
public class HazelcastAdditionalSettings
        implements BeanCreatedEventListener<HazelcastMemberConfiguration> {

    public HazelcastMemberConfiguration onCreated(BeanCreatedEvent<HazelcastMemberConfiguration> event) {
        HazelcastMemberConfiguration configuration = event.getBean();
        configuration.getGroupConfig().setName("micronaut-cluster");
        JoinConfig joinConfig = configuration.getNetworkConfig().getJoin();
        joinConfig.getMulticastConfig().setEnabled(false);
        joinConfig.getKubernetesConfig().setEnabled(true);
        return configuration;
    }
}
----

> We also create the following file to deploy the application into kubernetes.

Author: The file's name is kubernetes.yaml, is there a difference?

> We also create the following file to deploy the application into kubernetes.

Reviewer: I meant I do not see any "following" file here either yaml or yml. Did you refer to kubernetes.yaml in the next line or did you miss to include the file content here?

> We also create the following file to deploy the application into kubernetes.

Author: I changed it, and I found some bugs, I will push them now

----
This command will create a docker image named `caching-micronaut-microservices-on-kubernetes:0.1.0`. If you want an image with your choice of name to push it to some registry, you can run the following instead:

----
$ mvn clean compile com.google.cloud.tools:jib-maven-plugin:1.8.0:dockerBuild  -Dimage=YOUR-NAME/caching-micronaut-microservices-on-kubernetes:0.1.0
----

If you run the above command make sure to change the following part in the `kubernetes.yaml` file.
[source]
----
...
      containers:
        - name: hazelcast-micronaut-container
          image: caching-micronaut-microservices-on-kubernetes:0.1.0
          imagePullPolicy: IfNotPresent
          ports:
....
----

== Configure RBAC

https://github.com/hazelcast/hazelcast-kubernetes[Hazelcast Kubernetes discovery plugin] makes calls to Kubernetes API to provide automatic member discovery. Therefore, it needs to have specific ClusterRole rules granted. You can apply the minimal RBAC configuration (for the `default` service account in the `default` namespace) with the following command.

By this time make sure you have your kubernetes environment running.

> [source,shell script]

Author: Not really, I will change it to bash.

```suggestion
[source,bash]
```

----
kubectl apply -f rbac.yaml
----

> Note that:
> 
> - If you use service account other than `default` or namespace other than `default`, you need to modify `rbac.yaml` file.
> - If your Kubernetes cluster does not use RBAC, you can skip the "Configure RBAC" step

Reviewer: Let's make this note seciton cool (see the result [here](https://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#more-delimited-blocks)):
```suggestion
[NOTE]
====
- If you use service account other than `default` or namespace other than `default`, you need to modify `rbac.yaml` file.
- If your Kubernetes cluster does not use RBAC, you can skip the "Configure RBAC" step
====
```

== Deploy Micronaut Application to Kubernetes

[source,shell script]
----
kubectl apply -f kubernetes.yaml
----

== Testing the Application
Launch a curl container inside kubernetes cluster. This will create an interactable shell.

[source,shell script]
----
$ kubectl run curl --rm --image=radial/busyboxplus:curl -i --tty
----

> Put a value to the cluster

Reviewer: nit
```suggestion
Put a value to Hazelcast cluster:
```

[source,shell script]
----
$ curl  "http://quarkus-service/hazelcast/put?key=1&value=2"
{"value":"2","podName":"hazelcast-embedded-hazelcast-micronaut-container-0"}
----

Get the value from cluster in a loop and see that it is retrieved from different Pod Names.

[source,shell script]
----
$ while true; do curl "http://quarkus-service/hazelcast/get?key=1"; sleep 2;echo; done
{"value":"2","podName":"hazelcast-micronaut-container-1"}
{"value":"2","podName":"hazelcast-micronaut-container-0"}
...
----

You can stop the command by `CTRL + C` and exit the shell by typing `exit`.

== Tearing Down the Application
To delete all Kubernetes resources you created, run the following commands.

[source,shell script]
----
$ kubectl delete -f kubernetes.yaml
$ kubectl delete -f rbac.yaml
----

== Summary

In this guide, we bootstrapped a Micronaut application and deployed it to Kubernetes. We pushed a key-value pair to one of the applications. Then we could get the key-value pair from both.

== See Also

> - https://github.com/hazelcast-guides/caching-springboot-microservices-on-kubernetes[Caching SpringBoot Microservices with Hazelcast in Kubernetes]

Reviewer: This guide is outdated. Use the new one instead:
```suggestion
- xref:hazelcast-embedded-kubernetes:ROOT:index.adoc[SpringBoot Application with Embedded Hazelcast on Kubernetes]
```

> - https://github.com/hazelcast-guides/caching-micronaut-microservices-on-kubernetes[Caching Micronaut microservices on Kubernetes using Hazelcast]

Reviewer: This is the current one already.
```suggestion
```

> - https://github.com/hazelcast-guides/hazelcast-embedded-springboot[Hazelcast Embedded Spring Boot]

Reviewer: Use relative links instead
```suggestion
- xref:hazelcast-embedded-springboot:ROOT:index.adoc[Getting Started with Hazelcast using Spring Boot]
```
